crates/stages/src/stages/index_account_history.rs

```diff
@@ -0,0 +1,445 @@
+use crate::{
+    db::Transaction, ExecInput, ExecOutput, Stage, StageError, StageId, UnwindInput, UnwindOutput,
+};
+use itertools::Itertools;
+use reth_db::{
+    cursor::{DbCursorRO, DbCursorRW},
+    database::Database,
+    models::ShardedKey,
+    tables,
+    transaction::{DbTx, DbTxMut},
+    TransitionList,
+};
+use reth_primitives::{Address, TransitionId};
+use std::{collections::BTreeMap, fmt::Debug};
+use tracing::*;
+
+const INDEX_ACCOUNT_HISTORY: StageId = StageId("IndexAccountHistoryStage");
+
+const NUM_OF_INDICES_IN_SHARD: usize = 100;
+/// Account hashing stage hashes plain account.
+/// This is preparation before generating intermediate hashes and calculating Merkle tree root.
+#[derive(Debug)]
+pub struct IndexAccountHistoryStage {
+    /// Number of blocks after which the control
+    /// flow will be returned to the pipeline for commit.
+    pub commit_threshold: u64,
+}
+
+impl Default for IndexAccountHistoryStage {
+    fn default() -> Self {
+        Self { commit_threshold: 100_000 }
+    }
+}
+
+#[async_trait::async_trait]
+impl<DB: Database> Stage<DB> for IndexAccountHistoryStage {
+    /// Return the id of the stage
+    fn id(&self) -> StageId {
+        INDEX_ACCOUNT_HISTORY
+    }
+
+    /// Execute the stage.
+    async fn execute(
+        &mut self,
+        tx: &mut Transaction<'_, DB>,
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError> {
+        let stage_progress = input.stage_progress.unwrap_or_default();
+        let previous_stage_progress = input.previous_stage_progress();
+
+        // read account changeset, merge it into one changeset and calculate account hashes.
+        let from_transition = tx.get_block_transition(stage_progress)?;
+        // NOTE: can probably done more probabilistic take of bundles with transition but it is
+        // guess game for later. Transitions better reflect amount of work.
+        let to_block =
+            std::cmp::min(stage_progress + self.commit_threshold, previous_stage_progress);
```

Comment:
this can be done with `exec_or_return` macro (see other stages)

Reply:
I didn't want to use macro it hides this simple intention, and as `NOTE` said it would be  better to do bundles with transition id than a block number, so I did it like this.